services/sharedchannel/msg.go

```diff
@@ -6,48 +6,104 @@ package sharedchannel
 import (
 	"encoding/json"
 
+	"github.com/mattermost/mattermost-server/v5/mlog"
 	"github.com/mattermost/mattermost-server/v5/model"
 )
 
 // msgCache caches the work of converting a change in the Posts table to a remote cluster message.
 // Maps Post id to syncMsg.
 type msgCache map[string]syncMsg
 
-// syncMsg represents a change in content (post add/edit/delete, reaction add/remove).
+// syncMsg represents a change in content (post add/edit/delete, reaction add/remove, users).
 // It is sent to remote clusters as the payload of a `RemoteClusterMsg`.
 type syncMsg struct {
+	ChannelId string            `json:"channel_id"`
+	PostId    string            `json:"post_id"`
 	Post      *model.Post       `json:"post"`
+	Users     []*model.User     `json:"users"`
 	Reactions []*model.Reaction `json:"reactions"`
 }
 
-// postsToMsg takes a slice of posts and converts to a `RemoteClusterMsg` which can be
-// sent to a remote cluster
-func (scs *Service) postsToMsg(posts []*model.Post, cache msgCache) (model.RemoteClusterMsg, error) {
+func (sm syncMsg) ToJSON() ([]byte, error) {
+	b, err := json.Marshal(sm)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (sm syncMsg) String() string {
+	json, err := sm.ToJSON()
+	if err != nil {
+		return ""
+	}
+	return string(json)
+}
 
+// postsToMsg takes a slice of posts and converts to a `RemoteClusterMsg` which can be
+// sent to a remote cluster.
+func (scs *Service) postsToMsg(posts []*model.Post, cache msgCache, rc *model.RemoteCluster, lastSyncAt int64) (model.RemoteClusterMsg, error) {
 	syncMessages := make([]syncMsg, 0, len(posts))
 
 	for _, p := range posts {
+		if p.IsSystemMessage() { // don't sync system messages
+			continue
+		}
+
 		if sm, ok := cache[p.Id]; ok {
 			syncMessages = append(syncMessages, sm)
 			continue
 		}
 
-		reactions, err := scs.server.GetStore().Reaction().GetForPost(p.Id, true)
+		// any reactions originating from the remote cluster are filtered out
+		reactions, err := scs.server.GetStore().Reaction().GetForPostSince(p.Id, lastSyncAt, rc.RemoteId, true)
 		if err != nil {
 			return model.RemoteClusterMsg{}, err
 		}
 
+		postSync := p
+
+		// TODO:  don't include the post if only the reactions changed. Unfortunately there is no way to reliably know the
+		//        difference between an existing (synchronized) post with new reaction, and a brand new post (un-synchronized)
+		//        with a reaction.
+		//if p.EditAt < p.UpdateAt && p.CreateAt < p.UpdateAt && p.DeleteAt == 0 {
+		//	postSync = nil
+		//}
```

Comment:
Yes.  Currently the logic is assume everything needs to be sent unless we can detect it doesn't need to.  I need to re-visit this as it wasn't working as expected when I wrote it, but I have fixed other issues since then which might have been the cause.  I have another ticket to fix the redundant post sends.